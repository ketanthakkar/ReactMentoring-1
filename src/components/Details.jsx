import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const DetailsWrapper = styled.div`
    display: flex;
    position: relative;
    align-items: flex-start;
    justify-content: flex-start;
    font-size: 18px;
`;

const DetailsSearchButton = styled.button`
    position: absolute;
    right: 0;
    top: -20px;
    border: none;
    border-radius: 3px;
    background-color: #fff;
    color: #f55263;
    text-transform: uppercase;
    font-weight: bold;
    padding: 10px 30px;
    cursor: pointer;
    display: block;
    z-index: 1;
    
    &:hover {
        color: #fff;
        background-color: #f55263;
    }
`;

const DetailsImage = styled.img`
    height: 400px;
    padding-top: 30px;
    margin: 0 50px 50px 0;
`;

const DetailsDescription = styled.div`
    color: #ffffff;
`;

const DetailsTitle = styled.div`
    font-size: 40px;
    font-weight: bold;
    color: #f55263;
    margin-top: 40px;
    
    &::after {
        content: '${props => props.vote_average}';
        display: inline-block;
        position: absolute;
        margin-left: 10px;
        font-size: 20px;
        color: #84ba67;
        border: 1px solid #84ba67;
        border-radius: 28px;
        width: 40px;
        height: 40px;
        line-height: 40px;
        text-align: center;
    }
`;

const DetailsTagline = styled.div`
    color: #bdbdbd;
    margin: 15px 0;
`;

const DetailsYear = styled.span`
    font-size: 24px;
    margin-right: 25px;
`;

const DetailsRuntime = styled.span`
    font-size: 24px;
`;

const DetailsOverview = styled.div`
    color: #bdbdbd;
    font-size: 22px;
    margin: 20px 0;
`;

class Details extends React.Component {
    constructor(props) {
        super(props)
    }

    _goToSearchPage() {
        this.props.goToSearchPage();
    }

    render() {
        const {
            item
        } = this.props;

        const {
            poster_path,
            title,
            vote_average,
            tagline,
            release_date,
            runtime,
            overview
        } = item;

        return (
            <DetailsWrapper>
                <DetailsSearchButton onClick={this._goToSearchPage.bind(this)}>
                    search
                </DetailsSearchButton>
                <DetailsImage
                    src={poster_path}
                />
                <DetailsDescription>
                    <DetailsTitle vote_average={vote_average}>{title}</DetailsTitle>
                    <DetailsTagline>{tagline}</DetailsTagline>
                    <div>
                        <DetailsYear>
                            {release_date && release_date.slice(0, release_date.indexOf('-'))}
                        </DetailsYear>
                        <DetailsRuntime>{runtime} min</DetailsRuntime>
                    </div>
                    <DetailsOverview>{overview}</DetailsOverview>
                </DetailsDescription>
            </DetailsWrapper>
        )
    }
}

Details.propTypes = {
    item: PropTypes.object,
    poster_path: PropTypes.string,
    title: PropTypes.string,
    vote_average: PropTypes.number,
    tagline: PropTypes.string,
    release_date: PropTypes.string,
    runtime: PropTypes.number,
    overview: PropTypes.string,
    goToSearchPage: PropTypes.func
};

Details.defaultProps = {
    item: {},
    poster_path: '',
    title: '',
    vote_average: 0,
    tagline: '',
    release_date: '',
    runtime: null,
    overview: '',
    goToSearchPage: null
};

export default Details;