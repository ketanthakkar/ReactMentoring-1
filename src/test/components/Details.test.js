import React from 'react';
import { shallow } from 'enzyme';

import Details from 'components/Details';

describe('Details', () => {
    const mockGetItems = jest.fn();
    const details = shallow(<Details getItems={mockGetItems} />);

    it('should be like snapshot', () => {
        expect(details).toMatchSnapshot();
    });

    it('should call getItems function on DetailsSearchButton click', () => {
        details.find('Details__DetailsSearchButton').simulate('click');
        expect(mockGetItems.mock.calls.length).toBe(1);
    });
});
