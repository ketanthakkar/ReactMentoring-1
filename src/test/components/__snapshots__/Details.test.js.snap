// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details should be like snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Details
    getItem={
      [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
      }
    }
    item={Object {}}
    match={
      Object {
        "params": Object {
          "id": "",
        },
      }
    }
    searchBy=""
    sortBy=""
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Link
          replace={false}
          to="/search?search=&searchBy=&sortBy="
        >
          <Details__DetailsSearchButton>
            search
          </Details__DetailsSearchButton>
        </Link>,
        <Details__DetailsImage
          src={undefined}
        />,
        <Details__DetailsDescription>
          <Details__DetailsTitle
            vote_average={undefined}
          />
          <Details__DetailsTagline />
          <div>
            <Details__DetailsYear />
            <Details__DetailsRuntime>
               min
            </Details__DetailsRuntime>
          </div>
          <Details__DetailsOverview />
        </Details__DetailsDescription>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Details__DetailsSearchButton>
            search
          </Details__DetailsSearchButton>,
          "replace": false,
          "to": "/search?search=&searchBy=&sortBy=",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "search",
          },
          "ref": null,
          "rendered": "search",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "src": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Details__DetailsTitle
              vote_average={undefined}
            />,
            <Details__DetailsTagline />,
            <div>
              <Details__DetailsYear />
              <Details__DetailsRuntime>
                 min
              </Details__DetailsRuntime>
            </div>,
            <Details__DetailsOverview />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": undefined,
              "vote_average": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Details__DetailsYear />,
                <Details__DetailsRuntime>
                   min
                </Details__DetailsRuntime>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    undefined,
                    " min",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  undefined,
                  " min",
                ],
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Link
            replace={false}
            to="/search?search=&searchBy=&sortBy="
          >
            <Details__DetailsSearchButton>
              search
            </Details__DetailsSearchButton>
          </Link>,
          <Details__DetailsImage
            src={undefined}
          />,
          <Details__DetailsDescription>
            <Details__DetailsTitle
              vote_average={undefined}
            />
            <Details__DetailsTagline />
            <div>
              <Details__DetailsYear />
              <Details__DetailsRuntime>
                 min
              </Details__DetailsRuntime>
            </div>
            <Details__DetailsOverview />
          </Details__DetailsDescription>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Details__DetailsSearchButton>
              search
            </Details__DetailsSearchButton>,
            "replace": false,
            "to": "/search?search=&searchBy=&sortBy=",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "search",
            },
            "ref": null,
            "rendered": "search",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "src": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Details__DetailsTitle
                vote_average={undefined}
              />,
              <Details__DetailsTagline />,
              <div>
                <Details__DetailsYear />
                <Details__DetailsRuntime>
                   min
                </Details__DetailsRuntime>
              </div>,
              <Details__DetailsOverview />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": undefined,
                "vote_average": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Details__DetailsYear />,
                  <Details__DetailsRuntime>
                     min
                  </Details__DetailsRuntime>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": undefined,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": Array [
                      undefined,
                      " min",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    undefined,
                    " min",
                  ],
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
